### PR TITLE
[mpl_toolkits] Allow "figure" kwarg for host functions in parasite_axes

### PR DESCRIPTION
**Pull request by productivememberofsociety666**
Currently, ``host_axes`` and ``host_subplot`` of  ``mpl_toolkits.axes_grid1`` work on ``pyplot`` only. This patch allows a kwarg ``figure`` (defaulting to ``pyplot.gcf()``) to be passed so it becomes possible to use them on arbitrary figures as well.

### CODE REVIEW

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 1]
```diff
@@ -487,7 +487,10 @@ def host_axes(*args, **kwargs):
     import matplotlib.pyplot as plt
     axes_class = kwargs.pop("axes_class", None)
     host_axes_class = host_axes_class_factory(axes_class)
-    fig = plt.gcf()
+    if kwargs.get("figure") != None:
```
Comment (WeatherGod) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py]: don't do equality comparison to `None`. Always do `is` and `is not`.

**Comment by WeatherGod** [comment]
Also needs documentation (although, I am not clear where...). Besides the equality comment, this makes sense to me. Don't know why it wasn't coded that way before...

**Comment by productivememberofsociety666** [comment]
@WeatherGod thanks for your correction. Both of these functions weren't documented (except for their usage in examples) to begin with, so I'm not sure either. The whole AxesGrid toolkit seems a bit neglected to me, do people still use it?

**Comment by WeatherGod** [comment]
Oh, absolutely it gets used. It is one of the more popular toolkits.

**Comment by dopplershift** [comment]
Someone else just contributed some docs: #4864

**Comment by WeatherGod** [comment]
In fact, you might be interested in #4864 (just submitted) where someone gives some love to the documentation.

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 2]
```diff
@@ -484,20 +484,50 @@ def host_subplot_class_factory(axes_class):
...
+        `pyplot.gcf()`.
+
+    *args, **kwargs :
+        Will be passed on to the underlying ``Axes`` object creation.
+    """
     import matplotlib.pyplot as plt
     axes_class = kwargs.pop("axes_class", None)
     host_axes_class = host_axes_class_factory(axes_class)
-    fig = plt.gcf()
+    if kwargs.get("figure") is not None:
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py]: I would do this as 

```python
fig = kwargs.get('figure', None)
if fig is None:
    fig = plt.gcf()
```

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 3]
```diff
@@ -484,20 +484,50 @@ def host_subplot_class_factory(axes_class):
...
+        figure `pyplot.gcf()`.
+
+    *args, **kwargs :
+        Will be passed on to the underlying ``Axes`` object creation.
+    """
     import matplotlib.pyplot as plt
     axes_class = kwargs.pop("axes_class", None)
     host_subplot_class = host_subplot_class_factory(axes_class)
-    fig = plt.gcf()
+    if kwargs.get("figure") is not None:
```
Comment (tacaswell) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py]: ditto for the pattern above

**Comment by tacaswell** [comment]
Sorry this languished, in the future leave a comment after you finish with edits.  We don't get notifications on commit pushes, just comments.

**Comment by tacaswell** [comment]
This isn't perfect, but does not make it worse.

**Comment by FilipDominec** [comment]
Could you, please, point me to a minimum example of using `host_axes` once 
```python
self.fig = matplotlib.figure.Figure(figsize=(8,8))
```
is defined?

**Comment by smheidrich** [comment]
@FilipDominec 
````python
hax = host_axes([0.1,0.1,0.8,0.8], figure=fig)
hax.plot([0,1,2,3],[1,2,4,8])
fig.savefig("some_file.png")
````
Result looks very broken though, even when I set ``figure=plt.gcf()`` (which is weird because nothing else happens inside the function when ``figure`` is ``None``).

I'm using ``host_subplot`` with ``figure`` kwarg in my own code. I think I just added the kwarg for ``host_axes`` while I was at it because it was almost identical to ``host_subplot``.

**Comment by FilipDominec** [comment]
@smheidrich Thanks!

**Pull request merged**